lib/output_descriptor.ex

```diff
@@ -0,0 +1,527 @@
+    {:ok, script} = Script.create_p2pk(Point.sec(key))
+    script
+  end
+
+  def to_script(%__MODULE__{type: :pkh, key: key}) do
+    {:ok, script} = Script.public_key_to_p2pkh(key)
+    script
+  end
+
+  def to_script(%__MODULE__{type: :wpkh, key: key}) do
+    {:ok, script} = Script.public_key_to_p2wpkh(key)
+    script
+  end
+
+  def to_script(%__MODULE__{type: :sh, sub_descriptor: sub_descriptor})
+      when not is_nil(sub_descriptor) do
+    redeem_script = to_script(sub_descriptor)
+    {:ok, script} = Script.to_p2sh(redeem_script)
+    script
+  end
+
+  def to_script(%__MODULE__{type: :sh, key: key}) when is_binary(key) do
+    {:ok, script} = Script.create_p2sh(key)
+    script
+  end
+
+  def to_script(%__MODULE__{type: :wsh, sub_descriptor: sub_descriptor})
+      when not is_nil(sub_descriptor) do
+    witness_script = to_script(sub_descriptor)
+    {:ok, script} = Script.to_p2wsh(witness_script)
+    script
+  end
+
+  def to_script(%__MODULE__{type: :wsh, key: key}) when is_binary(key) do
+    {:ok, script} = Script.create_p2wsh(key)
+    script
+  end
+
+  def to_script(%__MODULE__{type: :multi, threshold: m, keys: keys}) do
+    {:ok, script} = Script.create_multi(m, keys)
+    script
+  end
+
+  def to_script(%__MODULE__{type: :sortedmulti, threshold: m, keys: keys}) do
+    sorted_keys =
+      Enum.sort(keys, fn a, b ->
+        a_sec = Point.sec(a)
+        b_sec = Point.sec(b)
+        a_sec <= b_sec
+      end)
+
+    {:ok, script} = Script.create_multi(m, sorted_keys)
+    script
+  end
+
+  def to_script(%__MODULE__{type: :addr, address: address}) do
+    {:ok, script, _network} = Script.from_address(address)
+    script
+  end
+
+  def to_script(%__MODULE__{type: :raw, raw_script: script_binary}) do
+    {:ok, script} = Script.parse_script(script_binary)
+    script
+  end
+
+  # Taproot output script generation
+  def to_script(%__MODULE__{type: :tr, key: key}) do
+    # Convert to an x-only public key if it's a Point
+    xonly_key =
+      case key do
+        %Point{} ->
+          # Extract the x coordinate only
+          serialized = Point.sec(key)
+          # Remove the first byte (compression marker) to get the x-only key
+          binary_part(serialized, 1, 32)
+
+        binary when is_binary(binary) and byte_size(binary) == 33 ->
+          # Remove first byte from compressed pubkey
+          binary_part(binary, 1, 32)
+
+        binary when is_binary(binary) and byte_size(binary) == 32 ->
+          # Already an x-only key
+          binary
+
+        _ ->
+          raise "Invalid key for Taproot descriptor"
+      end
+
+    # Create a pay-to-taproot script: OP_1 <32-byte x-only pubkey>
+    {:ok, script} = Script.create_p2tr(xonly_key)
+    script
+  end
+
+  @doc """
+  Converts an output descriptor to an address.
+  """
+  @spec to_address(t(), atom()) :: {:ok, String.t()} | {:error, String.t()}
+  def to_address(descriptor, network \\ :mainnet) do
```

Comment:
just my opinion, but i'd prefer avoiding a default value for network, like in `Script.to_address`